### DISCUSSION TITLE
moon vs turbo

### DISCUSSION THREAD

**Discussion by alexn-s** [moonrepo/moon#175] [Q&A]
i just found moon and it looks really promising with a ton of features.

currently i am using multiple monorepos with pnpm and turbo which works really well for me.
i am interested in how moon compares to turbo or lerna for example and what the experiences of other devs using moon are.

looking forward for moon to become the next monorepo manager 💖

**Comment by milesj** [comment]
Thanks for starting this. It's on my to do list, but definitely will require some deep investigative work. I would love to hear others thoughts on this matter.

I've never used Turborepo myself, so I can't compare at this time. The development of moon actually started before Turborepo was acquired by Vercel and released, so any similarities are coincidental. But in the grand scheme of things, a "build system" is typically a directed acyclic graph that is processed in topological order. The same could be said for Nx, Bazel, etc.

With that said, I have used Nx extensively, and plan on writing an in-depth comparison article in the future. Right now, we're trying to lock down the initial feature set for moon and stabilize it.

**Comment by milesj** [comment]
I've put together this comparison doc: https://moonrepo.dev/docs/comparison It's very high-level, but a good place to start. I will get more in-depth in the future once I've had time to investigate things further.

**Reply by jase88** [reply]
Thanks for comparing, looking great. 

Maybe adding an entry to https://monorepo.tools would be an option: https://github.com/nrwl/monorepo.tools

**Reply by milesj** [reply]
@jase88 Yeah it's on my to do also! It's a great resource.

**Comment by milesj** [comment]
Finally got around to writing this: https://moonrepo.dev/docs/comparison#turborepo

**Comment by klagrida** [comment]
Hi @milesj, I have two question about moon vs nx comparison. Can moon now or in the future:
- extend the graph with deps from another language like java via gradle or maven
- add a dynamic output for a task, for example for maven the output for a build is:`{m2RepoPath}/{projectGroupId}/{projectName}/{projectVersion}`

Both this features are covered by nx with `createNodes` and `createDependencies` functions.
Thanks

**Reply by milesj** [reply]
Hey 👋 

> extend the graph with deps from another language like java via gradle or maven

Do you have an example of what this looks like exactly?

> add a dynamic output for a task

In the next release, outputs can be defined using token vars & env vars. That's the extent of the dynamic support at the moment.

**Reply by klagrida** [reply]
I am referring to this [Extending the Project Graph of Nx](https://nx.dev/extending-nx/recipes/project-graph-plugins).
I made a plugin for maven, the idea was to parse `pom.xml` files and complete the graph with maven projects. Was asking if we can do the same with Moon.

**Reply by milesj** [reply]
At this time, there is not. 

We plan to support toolchain plugins hopefully by EOY. This will allow users to create their own "go", "ruby", "zig", etc integration.

We also plan to update extension plugins (https://moonrepo.dev/docs/guides/extensions) in the future, to do just that, extend things. Right now they are very basic.

**Comment by muuvmuuv** [comment]
Really want to also add my two cents here, the project sync command is just as awesome as it sounds, it syncs all relevant relations between TypeScript projects e.g. It builds your tsconfig paths, references and adds the correct package definition to the lib/app package.json dependencies based on your package manager. No magic on-the-fly types generation and weird looking tsconfigs.

Its fast and just works. We are never touching our package.json or tsconfig anymore only say who `dependsOn` what ^^. 

Task inheritance and task definitions for a tag or system is also very very awesome. Reduces a lot of clutter!

![image](https://github.com/user-attachments/assets/98861ab1-63c4-4a63-b2c7-90d9d8af9192)

Our action-graph: 
![SCR-20250327-osur](https://github.com/user-attachments/assets/30c4a7da-91a1-4f01-ba52-a3645823c6bf)
